File: spring-cloud-starter/src/main/java/com/jmsoftware/maf/springcloudstarter/helper/IpHelper.java
package com.jmsoftware.maf.springcloudstarter.helper;

import cn.hutool.core.util.StrUtil;
import cn.hutool.http.HttpUtil;
import lombok.*;
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.web.context.WebServerInitializedEvent;
import org.springframework.context.ApplicationListener;
import org.springframework.core.env.Environment;
import org.springframework.stereotype.Component;

import java.net.*;
import java.util.Enumeration;

/**
 * <h1>ServerConfiguration</h1>
 * <p>
 * Change description here.
 *
 * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com
 * @date 2019-04-26 16:02
 **/
@Slf4j
@Getter
@Component
@RequiredArgsConstructor
public class IpHelper implements ApplicationListener<WebServerInitializedEvent> {
    private static final String DEVELOPMENT_ENVIRONMENT = "development";
    @Getter(AccessLevel.NONE)
    private final Environment environment;
    private int serverPort;

    @Override
    public void onApplicationEvent(WebServerInitializedEvent event) {
        this.serverPort = event.getWebServer().getPort();
    }

    /**
     * Find public IP address.
     *
     * @return public IP
     */
    public String getPublicIp() {
        val jointProfiles = String.join(",", this.environment.getActiveProfiles());
        if (StrUtil.containsIgnoreCase(jointProfiles, DEVELOPMENT_ENVIRONMENT)) {
            log.debug("Current active profiles for environment contains: {}", DEVELOPMENT_ENVIRONMENT);
            return this.getInternetIp();
        }
        // An API provided by https://whatismyipaddress.com/api
        return HttpUtil.get("https://ipv4bot.whatismyipaddress.com/").trim();
    }

    /**
     * Get internet IP.
     *
     * @return internet IP
     */
    @SneakyThrows({SocketException.class})
    private String getInternetIp() {
        val intranetIp = this.getIntranetIp();
        val networks = NetworkInterface.getNetworkInterfaces();
        InetAddress ip;
        Enumeration<InetAddress> addresses;
        while (networks.hasMoreElements()) {
            addresses = networks.nextElement().getInetAddresses();
            while (addresses.hasMoreElements()) {
                ip = addresses.nextElement();
                if (ip instanceof Inet4Address
                        && ip.isSiteLocalAddress()
                        && !ip.getHostAddress().equals(intranetIp)) {
                    return ip.getHostAddress();
                }
            }
        }
        return intranetIp;
    }

    /**
     * Get intranet IP.
     *
     * @return intranet IP
     */
    @SneakyThrows({UnknownHostException.class})
    private String getIntranetIp() {
        return InetAddress.getLocalHost().getHostAddress();
    }
}

File: spring-cloud-starter/src/main/java/com/jmsoftware/maf/springcloudstarter/helper/IpHelper.kt
package com.jmsoftware.maf.springcloudstarter.helper

import cn.hutool.core.util.StrUtil
import cn.hutool.http.HttpUtil
import com.jmsoftware.maf.common.util.logger
import org.springframework.boot.web.context.WebServerInitializedEvent
import org.springframework.context.ApplicationListener
import org.springframework.core.env.Environment
import java.net.Inet4Address
import java.net.InetAddress
import java.net.NetworkInterface
import java.util.*

/**
 * <h1>ServerConfiguration</h1>
 *
 *
 * Change description here.
 *
 * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com
 * @date 2019-04-26 16:02
 */
class IpHelper(
    private val environment: Environment
) : ApplicationListener<WebServerInitializedEvent> {
    companion object {
        private const val DEVELOPMENT_ENVIRONMENT = "development"
        private val log = logger()
    }

    var serverPort = 0

    override fun onApplicationEvent(event: WebServerInitializedEvent) {
        serverPort = event.webServer.port
    }

    /**
     * Find public IP address.
     *
     * @return public IP
     */
    fun getPublicIp(): String {
        val jointProfiles = java.lang.String.join(",", *environment.activeProfiles)
        if (StrUtil.containsIgnoreCase(jointProfiles, DEVELOPMENT_ENVIRONMENT)) {
            log.debug("Current active profiles for environment contains: {}", DEVELOPMENT_ENVIRONMENT)
            return internetIp()
        }
        // An API provided by https://whatismyipaddress.com/api
        return try {
            HttpUtil.get("https://ipv4bot.whatismyipaddress.com/").trim { it <= ' ' }
        } catch (e: Exception) {
            log.warn(
                "Failed to get public IP address, fallback to intranet IP address. Exception: {}",
                e.message
            )
            internetIp()
        }
    }

    /**
     * Get internet IP.
     *
     * @return internet IP
     */
    private fun internetIp(): String {
        val intranetIp = intranetIp()
        val networks = NetworkInterface.getNetworkInterfaces()
        var ip: InetAddress
        var addresses: Enumeration<InetAddress>
        while (networks.hasMoreElements()) {
            addresses = networks.nextElement().inetAddresses
            while (addresses.hasMoreElements()) {
                ip = addresses.nextElement()
                if (ip is Inet4Address
                    && ip.isSiteLocalAddress()
                    && ip.getHostAddress() != intranetIp
                ) {
                    return ip.getHostAddress()
                }
            }
        }
        return intranetIp
    }

    /**
     * Get intranet IP.
     *
     * @return intranet IP
     */
    private fun intranetIp(): String = InetAddress.getLocalHost().hostAddress
}
